### DISCUSSION TITLE
Can I print to file on the City Hub?

### DISCUSSION THREAD

**Discussion by 200bolts** [pybricks/support#1627] [Q&A]
I have just followed and programmed a train using the Dec 7 video "How to make smart Lego Train layouts".  I have also added in a stopwatch to record the time between loops, or from green tile back to green tile.  Is it possible to write the time from the stopwatch to a file instead of the screen?  I am trying to map the speed over time until the battery is almost out of charge.  The run may be longer than an hour and I would rather not have the hub paired to a computer for that long.  I see in the documentation that in general the write function can write to a file, but I am not seeing how to setup the file, or how to pull the file back off the city hub.

**Comment by laurensvalk** [comment]
> Is it possible to write the time from the stopwatch to a file instead of the screen?

While there is no "file system", you can indeed write a few numbers to persistent storage. Are you using block-based coding or Python? I can put a few samples together.

> I am trying to map the speed over time until the battery is almost out of charge. The run may be longer than an hour and I would rather not have the hub paired to a computer for that long.

Since this is essentially a single number (we can override the previous value if the next is greater), this should work well despite not having a full log file.

For your application specifically though, you could also dynamically adjust the power level based on the battery so it keeps running at a fixed speed. For example by doing `duty = 50 * 7200 / hub.battery.voltage()`. Here `50` is the "power level", and it is scaled by a term based on the battery. If the battery voltage decreases, it increases the duty cycle value to keep your train running at roughly the same speed.

**Reply by 200bolts** [reply]
> While there is no "file system", you can indeed write a few numbers to persistent storage. Are you using block-based coding or Python? I can put a few samples together. ... Since this is essentially a single number (we can override the previous value if the next is greater), this should work well despite not having a full log file.

I start in block based and then switched to Python, as I do not believe this can be done with the blocks.  I am expecting between 500-700 data points.  Will I need to create a variable and keep appending the new stopwatch value to the end?  I would then write the variable data to the hub just prior to the shutdown?  Or I could write the string to persistent storage and then the next time it reaches the green tile, append the time to the string and overwrite with the new longer string.  The plan is to let the program run until the battery is mostly discharged, then have the program end.  I would then connect to a computer to extract the data.

> For your application specifically though, you could also dynamically adjust the power level based on the battery so it keeps running at a fixed speed. 

I will be keeping the power level the same as I am trying to create a graph to show the speed curve so I can compare one setup against another.

**Reply by laurensvalk** [reply]
> I do not believe this can be done with the blocks

You can call any Python function from your block code. So you could write one Python function to do the writing, but keep using blocks for everything else.

> I am expecting between 500-700 data points.

This won't fit. The user data is 128 bytes. It's more intended for keeping persistent settings, measurement calibration data, and so on.

What is the main issue with staying connected? When benchmarking a physical system with electromotors, the overhead of connectivity is comparatively much smaller.

**Reply by 200bolts** [reply]
> This won't fit. The user data is 128 bytes.

OK.  Thank you for this information.  

> What is the main issue with staying connected?

My computer is a desktop and not in the same room as the track.   If the city hub had the capability of storing the data, it would have been easier to move the hub to the computer than moving the computer to the hub.  I will keep searching for the solution to what I am trying to accomplish.  

Thank you for your help and information.

**Reply by BertLindeman** [reply]
Maybe a strange idea, but . . . 

Would you be able to 

- load the program in the hub from your PC

- stop the program and disconnect from the pc
- connect the hub to an Android phone or tablet 
- run your test (by just pressing the green button on the hub) 

The phone or tablet can be moved easily.

Bert
